### PR TITLE
Relocate the code for the ensureHashed Classes

### PR DESCRIPTION
**Pull request by ChengJin01**
The changes move the code specific to the ensureHashed Classes
to `EnsureHashedConfig.cpp` so as to make room for the code with
initialization at startup.

Signed-off-by: ChengJin01 <jincheng@ca.ibm.com>

### CODE REVIEW

**Comment by ChengJin01** [comment]
This is not related to #19714 but I'd like to relocate the code there to tide them up for the initialization related code at startup before make any changes in `jvminit.c` for #19714.

**Comment by ChengJin01** [comment]
The PR has been verified in personal builds on all supported platforms.

Reviewer: @keithc-ca
FYI: @tajila, @pshipton

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: runtime/vm/CMakeLists.txt [hunk 1]
```diff
@@ -98,6 +98,7 @@ set(main_sources
 	exceptiondescribe.c
 	exceptionsupport.c
 	extendedMessageNPE.cpp
+	EnsureHashedConfig.cpp
```
[the remark this answers is not in the document]
Comment (ChengJin01) [reply runtime/vm/CMakeLists.txt]: Fixed.

File: runtime/vm/EnsureHashedConfig.cpp [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+				0,
+				0,
+				OMRMEM_CATEGORY_VM,
+				ensureHashedHashFn,
+				ensureHashedHashEqualFn,
+				NULL,
+				PORTLIB);
+
+		if (NULL == jvm->ensureHashedClasses) {
+			/* we failed the allocation. */
```
[the remark this answers is not in the document]
Comment (ChengJin01) [reply runtime/vm/EnsureHashedConfig.cpp]: Fixed everywhere.

File: runtime/vm/EnsureHashedConfig.cpp [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+	hashtableEntry = (J9UTF8 *)j9mem_allocate_memory(classNameLength + sizeof(J9UTF8_LENGTH(hashtableEntry)), OMRMEM_CATEGORY_VM);
+	if (NULL == hashtableEntry) {
+		j9nls_printf(PORTLIB, J9NLS_ERROR, J9NLS_ENSUREHASHED_CONFIG_OUT_OF_MEM);
+		return JNI_ENOMEM;
+	}
+	memcpy(J9UTF8_DATA(hashtableEntry), className, classNameLength);
+	J9UTF8_SET_LENGTH(hashtableEntry, (U_16)classNameLength);
+
+	if (TRUE == isAdd) {
```
[the remark this answers is not in the document]
Comment (ChengJin01) [reply runtime/vm/EnsureHashedConfig.cpp]: Fixed.

File: runtime/vm/EnsureHashedConfig.cpp [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+ * @brief This function parses a specific option.
+ *
+ * @param option the option to be parsed
+ * @returns JNI_OK on success
+ */
+static UDATA
+parseEnsureHashedOption(J9JavaVM *jvm, char *option, BOOLEAN isAdd)
+{
+	PORT_ACCESS_FROM_JAVAVM(jvm);
+	char *className = option;
```
[the remark this answers is not in the document]
Comment (ChengJin01) [reply runtime/vm/EnsureHashedConfig.cpp]: Fixed.

File: runtime/vm/EnsureHashedConfig.cpp [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+	size_t classNameLength = strlen(className);
+	J9UTF8 *hashtableEntry;
+
+	if (0 == strcmp(option, "")) {
```
[the remark this answers is not in the document]
Comment (ChengJin01) [reply runtime/vm/EnsureHashedConfig.cpp]: Fixed.

File: runtime/vm/EnsureHashedConfig.cpp [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+		/* Empty class name. */
+		j9nls_printf(PORTLIB, J9NLS_ERROR, J9NLS_VM_INVALID_EMPTY_OPTION, isAdd ? VMOPT_XXENABLEENSUREHASHED : VMOPT_XXDISABLEENSUREHASHED);
+		return J9VMDLLMAIN_FAILED;
+	}
+
+	if (NULL == jvm->ensureHashedClasses) {
+		jvm->ensureHashedClasses = hashTableNew(OMRPORT_FROM_J9PORT(PORTLIB),
+				J9_GET_CALLSITE(),
+				ENSUREHASHED_INITIAL_TABLE_SIZE,
+				sizeof(char*),
```
[the remark this answers is not in the document]
Comment (ChengJin01) [reply runtime/vm/EnsureHashedConfig.cpp]: Fixed.

File: runtime/vm/EnsureHashedConfig.cpp [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+				PORTLIB);
+
+		if (NULL == jvm->ensureHashedClasses) {
+			/* we failed the allocation. */
+			j9nls_printf(PORTLIB, J9NLS_ERROR, J9NLS_ENSUREHASHED_CONFIG_OUT_OF_MEM);
+			return JNI_ENOMEM;
+		}
+	}
+
+	hashtableEntry = (J9UTF8 *)j9mem_allocate_memory(classNameLength + sizeof(J9UTF8_LENGTH(hashtableEntry)), OMRMEM_CATEGORY_VM);
```
[the remark this answers is not in the document]
Comment (ChengJin01) [reply runtime/vm/EnsureHashedConfig.cpp]: Fixed.

File: runtime/vm/EnsureHashedConfig.cpp [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+	}
+	memcpy(J9UTF8_DATA(hashtableEntry), className, classNameLength);
+	J9UTF8_SET_LENGTH(hashtableEntry, (U_16)classNameLength);
+
+	if (TRUE == isAdd) {
+		/* -XX:+EnsureHashed: */
+		if (NULL == hashTableFind(jvm->ensureHashedClasses, &hashtableEntry)) {
+			if (NULL == hashTableAdd(jvm->ensureHashedClasses, &hashtableEntry)) {
+				j9mem_free_memory(hashtableEntry);
+				hashtableEntry = NULL;
```
[the remark this answers is not in the document]
Comment (ChengJin01) [reply runtime/vm/EnsureHashedConfig.cpp]: Fixed.

File: runtime/vm/EnsureHashedConfig.cpp [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+parseEnsureHashedConfig(J9JavaVM *jvm, char *options, BOOLEAN isAdd)
+{
+	UDATA result = JNI_OK;
+	char *nextOption = NULL;
+	char *cursor = options;
+	PORT_ACCESS_FROM_JAVAVM(jvm);
+
+	/* parse out each of the options. */
+	while (NULL != strstr(cursor, ",")) {
+		nextOption = scan_to_delim(PORTLIB, &cursor, ',');
```
[the remark this answers is not in the document]
Comment (ChengJin01) [reply runtime/vm/EnsureHashedConfig.cpp]: Fixed.

File: runtime/vm/EnsureHashedConfig.cpp [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+		if (NULL == nextOption) {
+			result = JNI_ERR;
+			break;
+		} else {
+			result = parseEnsureHashedOption(jvm, nextOption, isAdd);
+			j9mem_free_memory(nextOption);
+			nextOption = NULL;
```
[the remark this answers is not in the document]
Comment (ChengJin01) [reply runtime/vm/EnsureHashedConfig.cpp]: Fixed.

**Comment by keithc-ca** [comment]
Jenkins test sanity amac jdk21

**Pull request merged**